File: app/plant_tracking.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from math import isfinite
from typing import Any, Mapping, Sequence


ALLOWED_STATUSES = frozenset({"active", "dead", "removed"})
ALLOWED_HEALTH = frozenset({"unknown", "good", "watch", "poor"})
ALLOWED_EVENT_KINDS = frozenset({"note", "health", "status"})


def _required(value: object, label: str) -> str:
    text = str(value or "").strip()
    if not text:
        raise ValueError(f"{label} is required")
    return text


def _optional_text(value: object) -> str:
    return str(value or "").strip()


def _iso_date(value: object, label: str, *, optional: bool = False) -> str:
    text = _optional_text(value)
    if optional and not text:
        return ""
    try:
        parsed = date.fromisoformat(text)
    except ValueError as exc:
        raise ValueError(f"{label} must be YYYY-MM-DD") from exc
    if parsed.isoformat() != text:
        raise ValueError(f"{label} must be YYYY-MM-DD")
    return text


def _coordinate(value: object, label: str) -> float | None:
    if value is None or value == "":
        return None
    if isinstance(value, bool):
        raise ValueError(f"{label} must be numeric")
    try:
        number = float(value)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"{label} must be numeric") from exc
    if not isfinite(number):
        raise ValueError(f"{label} must be finite")
    return number


@dataclass(frozen=True)
class PlantRecord:
    id: str
    field_id: str
    planting_batch_id: str = ""
    label: str = ""
    planted_date: str = ""
    variety: str = ""
    latitude: float | None = None
    longitude: float | None = None
    status: str = "active"
    health: str = "unknown"
    notes: str = ""

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "PlantRecord":
        return cls(
            id=_required(value.get("id"), "plant id"),
            field_id=_required(value.get("field_id"), "field id"),
            planting_batch_id=_optional_text(value.get("planting_batch_id")),
            label=_optional_text(value.get("label")),
            planted_date=_iso_date(value.get("planted_date"), "planted_date", optional=True),
            variety=_optional_text(value.get("variety")),
            latitude=_coordinate(value.get("latitude"), "latitude"),
            longitude=_coordinate(value.get("longitude"), "longitude"),
            status=_required(value.get("status") or "active", "plant status"),
            health=_required(value.get("health") or "unknown", "plant health"),
            notes=_optional_text(value.get("notes")),
        )

    def validate(self) -> None:
        _required(self.id, "plant id")
        _required(self.field_id, "field id")
        if self.status not in ALLOWED_STATUSES:
            raise ValueError(f"Unsupported plant status: {self.status}")
        if self.health not in ALLOWED_HEALTH:
            raise ValueError(f"Unsupported plant health: {self.health}")
        if self.planted_date:
            _iso_date(self.planted_date, "planted_date")
        if (self.latitude is None) != (self.longitude is None):
            raise ValueError("latitude and longitude must be supplied together")
        if self.latitude is not None:
            if not isfinite(self.latitude) or not -90 <= self.latitude <= 90:
                raise ValueError("latitude is outside the supported range")
            assert self.longitude is not None
            if not isfinite(self.longitude) or not -180 <= self.longitude <= 180:
                raise ValueError("longitude is outside the supported range")


@dataclass(frozen=True)
class PlantEvent:
    id: str
    plant_id: str
    event_date: str
    kind: str
    value: str = ""
    notes: str = ""

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "PlantEvent":
        return cls(
            id=_required(value.get("id"), "event id"),
            plant_id=_required(value.get("plant_id"), "plant id"),
            event_date=_iso_date(value.get("event_date"), "event_date"),
            kind=_required(value.get("kind"), "event kind"),
            value=_optional_text(value.get("value")),
            notes=_optional_text(value.get("notes")),
        )

    def validate(self) -> None:
        _required(self.id, "event id")
        _required(self.plant_id, "plant id")
        _iso_date(self.event_date, "event_date")
        if self.kind not in ALLOWED_EVENT_KINDS:
            raise ValueError(f"Unsupported plant event kind: {self.kind}")
        if self.kind == "health" and self.value not in ALLOWED_HEALTH:
            raise ValueError(f"Unsupported plant health: {self.value}")
        if self.kind == "status" and self.value not in ALLOWED_STATUSES:
            raise ValueError(f"Unsupported plant status: {self.value}")
        if self.kind == "note" and not self.value and not self.notes:
            raise ValueError("note event requires value or notes")


def project_plant(
    plant: PlantRecord,
    events: Sequence[PlantEvent],
) -> dict[str, object]:
    """Project current per-plant state from an optional plant record and immutable events."""

    plant.validate()
    seen: set[str] = set()
    ordered: list[PlantEvent] = []
    planted = date.fromisoformat(plant.planted_date) if plant.planted_date else None

    for event in events:
        event.validate()
        if event.plant_id != plant.id:
            raise ValueError("plant event belongs to another plant")
        if event.id in seen:
            raise ValueError(f"Duplicate plant event id: {event.id}")
        seen.add(event.id)
        if planted is not None and date.fromisoformat(event.event_date) < planted:
            raise ValueError("plant event cannot predate planted_date")
        ordered.append(event)

    ordered.sort(key=lambda item: (item.event_date, item.id))
    status = plant.status
    health = plant.health
    for event in ordered:
        if event.kind == "status":
            status = event.value
        elif event.kind == "health":
            health = event.value

    return {
        "plant_id": plant.id,
        "field_id": plant.field_id,
        "planting_batch_id": plant.planting_batch_id,
        "label": plant.label,
        "planted_date": plant.planted_date,
        "variety": plant.variety,
        "latitude": plant.latitude,
        "longitude": plant.longitude,
        "status": status,
        "health": health,
        "notes": plant.notes,
        "last_event_date": ordered[-1].event_date if ordered else "",
        "event_count": len(ordered),
    }

File: android/app/src/main/java/gr/mastixa/manager/PlantTracking.java
package gr.mastixa.manager;

import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.*;

/** Pure optional per-plant state projection. No database or Android runtime dependency. */
public final class PlantTracking {
    public static final Set<String> STATUSES = Set.of("active", "dead", "removed");
    public static final Set<String> HEALTH = Set.of("unknown", "good", "watch", "poor");
    public static final Set<String> EVENT_KINDS = Set.of("note", "health", "status");

    private PlantTracking() {}

    public record Plant(
        String id,
        String fieldId,
        String plantingBatchId,
        String label,
        String plantedDate,
        String variety,
        Double latitude,
        Double longitude,
        String status,
        String health,
        String notes
    ) {}

    public record Event(
        String id,
        String plantId,
        String eventDate,
        String kind,
        String value,
        String notes
    ) {}

    public record Snapshot(
        String plantId,
        String fieldId,
        String plantingBatchId,
        String label,
        String plantedDate,
        String variety,
        Double latitude,
        Double longitude,
        String status,
        String health,
        String notes,
        String lastEventDate,
        int eventCount
    ) {}

    private static String required(String value, String label) {
        String text = value == null ? "" : value.trim();
        if (text.isEmpty()) throw new IllegalArgumentException(label + " is required");
        return text;
    }

    private static String text(String value) { return value == null ? "" : value.trim(); }

    private static LocalDate isoDate(String value, String label, boolean optional) {
        String text = text(value);
        if (optional && text.isEmpty()) return null;
        try {
            LocalDate parsed = LocalDate.parse(text);
            if (!parsed.toString().equals(text)) throw new DateTimeParseException("non-canonical", text, 0);
            return parsed;
        } catch (DateTimeParseException error) {
            throw new IllegalArgumentException(label + " must be YYYY-MM-DD", error);
        }
    }

    private static void validatePlant(Plant plant) {
        Objects.requireNonNull(plant, "plant");
        required(plant.id(), "plant id");
        required(plant.fieldId(), "field id");
        if (!STATUSES.contains(plant.status()))
            throw new IllegalArgumentException("Unsupported plant status: " + plant.status());
        if (!HEALTH.contains(plant.health()))
            throw new IllegalArgumentException("Unsupported plant health: " + plant.health());
        isoDate(plant.plantedDate(), "planted_date", true);
        if ((plant.latitude() == null) != (plant.longitude() == null))
            throw new IllegalArgumentException("latitude and longitude must be supplied together");
        if (plant.latitude() != null) {
            if (!Double.isFinite(plant.latitude()) || plant.latitude() < -90 || plant.latitude() > 90)
                throw new IllegalArgumentException("latitude is outside the supported range");
            if (!Double.isFinite(plant.longitude()) || plant.longitude() < -180 || plant.longitude() > 180)
                throw new IllegalArgumentException("longitude is outside the supported range");
        }
    }

    private static void validateEvent(Event event) {
        Objects.requireNonNull(event, "event");
        required(event.id(), "event id");
        required(event.plantId(), "plant id");
        isoDate(event.eventDate(), "event_date", false);
        if (!EVENT_KINDS.contains(event.kind()))
            throw new IllegalArgumentException("Unsupported plant event kind: " + event.kind());
        String value = text(event.value());
        if (event.kind().equals("health") && !HEALTH.contains(value))
            throw new IllegalArgumentException("Unsupported plant health: " + value);
        if (event.kind().equals("status") && !STATUSES.contains(value))
            throw new IllegalArgumentException("Unsupported plant status: " + value);
        if (event.kind().equals("note") && value.isEmpty() && text(event.notes()).isEmpty())
            throw new IllegalArgumentException("note event requires value or notes");
    }

    public static Snapshot project(Plant plant, List<Event> events) {
        validatePlant(plant);
        Objects.requireNonNull(events, "events");
        LocalDate planted = isoDate(plant.plantedDate(), "planted_date", true);
        Set<String> seen = new HashSet<>();
        List<Event> ordered = new ArrayList<>();
        for (Event event : events) {
            validateEvent(event);
            if (!event.plantId().equals(plant.id()))
                throw new IllegalArgumentException("plant event belongs to another plant");
            if (!seen.add(event.id()))
                throw new IllegalArgumentException("Duplicate plant event id: " + event.id());
            LocalDate eventDate = isoDate(event.eventDate(), "event_date", false);
            if (planted != null && eventDate.isBefore(planted))
                throw new IllegalArgumentException("plant event cannot predate planted_date");
            ordered.add(event);
        }
        ordered.sort(Comparator.comparing(Event::eventDate).thenComparing(Event::id));

        String status = plant.status();
        String health = plant.health();
        for (Event event : ordered) {
            if (event.kind().equals("status")) status = text(event.value());
            else if (event.kind().equals("health")) health = text(event.value());
        }

        return new Snapshot(
            plant.id(),
            plant.fieldId(),
            text(plant.plantingBatchId()),
            text(plant.label()),
            text(plant.plantedDate()),
            text(plant.variety()),
            plant.latitude(),
            plant.longitude(),
            status,
            health,
            text(plant.notes()),
            ordered.isEmpty() ? "" : ordered.get(ordered.size() - 1).eventDate(),
            ordered.size()
        );
    }
}
